src/Heap.sol

```diff
@@ -32,209 +36,199 @@ library BasicHeap {
-    /// @dev Only call this function with ranks within array's bounds.
-    /// @param _heap The heap to modify.
-    /// @param _rank1 The rank of the first account in the heap.
-    /// @param _rank2 The rank of the second account in the heap.
-    function swap(
-        Heap storage _heap,
-        uint256 _rank1,
-        uint256 _rank2
-    ) private {
-        Account memory accountOldRank1 = getAccount(_heap, _rank1);
-        Account memory accountOldRank2 = getAccount(_heap, _rank2);
-        setAccount(_heap, _rank1, accountOldRank2);
-        setAccount(_heap, _rank2, accountOldRank1);
+        setAccount(_heap, _accountToShift, _index);
     }
 
-    /// @notice Moves an account up the heap until its value is smaller than the one of its parent.
-    /// @dev This functions restores the invariant about the order of the values stored when the account at `_rank` is the only one with value greater than what it should be.
+    /// @notice Moves an account down the heap until its value is greater than the ones of its children.
+    /// @dev This functions restores the invariant about the order of the values stored when the account to move is the only one with value smaller than what it should be.
     /// @param _heap The heap to modify.
-    /// @param _rank The index of the account to move.
-    function shiftUp(Heap storage _heap, uint256 _rank) private {
-        Account memory initialAccount = getAccount(_heap, _rank);
-        uint256 initialValue = initialAccount.value;
-        while (_rank > 1 && initialValue > getAccount(_heap, _rank / 2).value) {
-            setAccount(_heap, _rank, getAccount(_heap, _rank / 2));
-            _rank /= 2;
+    /// @param _size The size of the heap.
+    /// @param _accountToShift The account to move.
+    /// @param _index The index of the account to move.
+    function shiftDown(Heap storage _heap, uint256 _size, Account memory _accountToShift, uint256 _index) private {
+        uint256 childIndex = (_index << 1) + 1;
+        uint256 rightChildIndex;
+        // At this point, childIndex (resp. childIndex+1) is the index of the left (resp. right) child.
+
+        while (childIndex < _size) {
+            Account memory childToSwap = _heap.accounts[childIndex];
+
+            // Find the child with largest value.
+            unchecked {
+                rightChildIndex = childIndex + 1; // This cannot overflow because childIndex < size.
+            }
+
+            if (rightChildIndex < _size) {
+                Account memory rightChild = _heap.accounts[rightChildIndex];
+                if (rightChild.value > childToSwap.value) {
+                    childToSwap = rightChild;
+                    childIndex = rightChildIndex;
+                }
+            }
+
+            if (childToSwap.value > _accountToShift.value) {
+                setAccount(_heap, childToSwap, _index);
+                _index = childIndex;
+                childIndex = (childIndex << 1) + 1;
```

Comment:
I believe this cannot overflow, which means that it would be safe to have this entire code block unchecked.

Reply:
Since you're both in favor, I'm wrapping it inside an unchecked block.